### PR TITLE
Add DRM sources to sfml-window

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

This moves the 3rd party DRM sources into SFML itself since we need to regularly modify these files. Note that we can't immediately start building this as C++ code since it uses C features or extensions that don't work in C++. That transition will require some extra work.

Addresses #2252 

While I was working on this I realized there was no cache variable for `SFML_USE_DRM` so I added that. This means you can edit the CMakeCache to enable or disable this setting without having to reconfigure the project.

How far do we want to take this? Do we want to refactor this as C++ code and do some modest style fixes in the same PR that moves the code or do we tackle this in smaller pieces?

EDT: I expanded the scope of this PR to also include all of the refactoring required to make this code match the rest of SFML.

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
Here's a sneak peak of the errors we get when trying to build this C code as C++.

```
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp: In function ‘void drm_fb_destroy_callback(gbm_bo*, void*)’:
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:46:29: error: invalid conversion from ‘void*’ to ‘drm_fb*’ [-fpermissive]
   46 |         struct drm_fb *fb = data;
      |                             ^~~~
      |                             |
      |                             void*
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp: In function ‘drm_fb* drm_fb_get_from_bo(gbm_bo*)’:
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:57:49: error: invalid conversion from ‘void*’ to ‘drm_fb*’ [-fpermissive]
   57 |         struct drm_fb *fb = gbm_bo_get_user_data(bo);
      |                             ~~~~~~~~~~~~~~~~~~~~^~~~
      |                                                 |
      |                                                 void*
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:66:20: error: invalid conversion from ‘void*’ to ‘drm_fb*’ [-fpermissive]
   66 |         fb = calloc(1, sizeof *fb);
      |              ~~~~~~^~~~~~~~~~~~~~~
      |                    |
      |                    void*
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:103:79: error: ISO C++ forbids compound-literals [-Werror=pedantic]
  103 |                 memcpy(handles, (uint32_t [4]){gbm_bo_get_handle(bo).u32,0,0,0}, 16);
      |                                                                               ^
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:103:47: error: taking address of temporary array
  103 |                 memcpy(handles, (uint32_t [4]){gbm_bo_get_handle(bo).u32,0,0,0}, 16);
      |                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:104:75: error: ISO C++ forbids compound-literals [-Werror=pedantic]
  104 |                 memcpy(strides, (uint32_t [4]){gbm_bo_get_stride(bo),0,0,0}, 16);
      |                                                                           ^
/home/thrasher/Projects/sfml/src/SFML/Window/DRM/Common.cpp:104:47: error: taking address of temporary array
  104 |                 memcpy(strides, (uint32_t [4]){gbm_bo_get_stride(bo),0,0,0}, 16);
      |                                               ^~~~~~~~~~~~~~~
```

Fixing the compound literals is easy if we could use C++11's `std::array` but will be a bit harder when restricted to C++03. We also need to do some style fixes like replacing all the tabs with spaces. Much of this won't be hard, we just need to agree on what exactly we want to change.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
I expanded the scope of this PR to also include all of the refactoring required to make this code match the rest of SFML.

**Comment by ChrisThrasher** [comment]
@substring @oomek Can you try this out and make sure I didn't break anything? Thanks!

**Comment by oomek** [comment]
@ChrisThrasher Sorry, I'm a bit preoccupied by the bathroom renovation, I'll check as soon as I can. Thank you very much for getting your hands dirty.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by substring** [comment]
We've stumbled upon some very rare cases where that part would fail
 https://github.com/SFML/SFML/blob/4134e358310cf200ec52fbda3fb7b83027e6c1e1/src/SFML/Window/DRM/DRMContext.cpp#L107-L118

Can you just keep L109->116 from that block ? Let's always trigger the `drmModeSetCrtc`, whichever the situation.

Beside that case, tests were good so far. I'll let @oomek comment further if he wishes

**Comment by oomek** [comment]
> Is this a bug I introduced in this PR?

Not sure if it's strictly a bug but an optimization added by @substring that does not play nicely when the drm app is launched from linux session in VirtualBox. It resulted in a black screen, no console visible upon exiting the app, so it's better to remove it until further investigation. Thanks for the commit reverting it, and your overall involvement in SFML-ifying the common drm functions. I had a chance to test it this evening and all seems to be working as expected.

**Review by oomek — APPROVED** [review]

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:159]: I tried to use `sf::Uint64` here instead but it failed to compile because `Uint64` didn't match the definition of `uint64_t`. I believe one of them was an `unsigned long` and the other an `unsigned long long`.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:149]: I'd prefer to replace this with `operator new` but I'm not sure if `new + memset` is better than `calloc + reinterpret_cast`. I seldom write code so low level so hopefully someone can help find a better way to express with within the limits of C++03.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:139]: Can we get around manually calling `free` when we don't have access to smart pointers? I couldn't think of a way to get rid of this easily.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:320]: Should we replace the use `printf` with `sf::err`?

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:281]: Is `fd` a well understood abbreviation? I'm assuming it means file descriptor but I'm not sure. Maybe we ought to use more expressive variable names.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:285]: I hope we can find a better name than `ret`. `ret` tells us nothing about what the variable actually means.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (Bromeon) [reply src/SFML/Window/DRM/DRMContext.cpp:149]: This should not be `reinterpret_cast`, anyway.
[`std::calloc()`](https://en.cppreference.com/w/cpp/memory/c/calloc) returns `void*`, which is convertible to other pointer types via `static_cast`.

But couldn't we just do this?
```cpp
fb = new DrmFb();
```
That type is defined as a POD:
https://github.com/SFML/SFML/blob/22033670f8800f077c2f4cd6419b4405feec1240/src/SFML/Window/DRM/DRMContext.hpp#L59-L63

All members (also PODs) will be zero-initialized.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (Bromeon) [reply src/SFML/Window/DRM/DRMContext.cpp:139]: It's probably OK as long as the semantics are easy to follow. Workarounds are likely more complex, and this would also allow relatively easy migration to `std::unique_ptr` in SFML 3 (especially when using `new/delete`).

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.cpp:149]: Are you sure they're zero-initialized? My impression was that they're left uninitialized.

Comment (Bromeon) [reply src/SFML/Window/DRM/DRMContext.cpp:149]: The subtle difference between:
```cpp
fb = new DrmFb; // uninit
fb = new DrmFb(); // zero
```
😉 

It also works for arrays:
```cpp
ptr = new int[5]; // uninit
ptr = new int[5](); // zero
```

Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.cpp:149]: https://godbolt.org/z/jYn3Kq398

That's cool, I didn't know that! In that case this is an easy fixup then.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.cpp:312]: Good find! I'll declare this variable where it's being initialized which also lets me mark it `const`.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:309]: I want to change `device` and `modeStr` to `std::string` but I'm not sure how to handle line 359.
```cpp
if (modeStr && *modeStr)
```
I suppose this is saying "if the string is not null and the first character is not null"? Is that the right way to read this? Would this translate to `if (!modeStr.empty()`?

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.cpp:309]: This `char*` return value is seemingly behaving like a `std::optional<std::string>`. There may be no string, or there may be an empty string. Those are distinct shapes. Because of that, I don't think we can replace this with merely a `std::string` since that would fail to model the distinction between a null string and an empty string.

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [inline src/SFML/Window/DRM/DRMContext.cpp:294]: I'm not a fan of using an environment variable to dynamic enable or disable debug logs. I want to change this to instead use the `SFML_DEBUG` preprocessor symbol. That way we stop doing so many environment variable lookups.

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: src/SFML/Window/DRM/DRMContext.hpp
Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.hpp]: Good idea! The diff to do this was tiny so I figured it was worth adding to this PR even if it's a little bit out of scope.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by Bambo-Borris — APPROVED** [review]
Feels like we're at diminishing returns style improvement wise. I can't see any obvious issues so I reckon it's ready

**Comment by ChrisThrasher** [comment]
@substring @oomek I made substantial changes since I last pinged you about this PR but have since finished everything I wanted to do. If possible I'd really appreciate you give this a review and test it out on your end so we can get this merged. I think this will help speed up all future work on DRM code. Thanks!

**Comment by oomek** [comment]
Just tested. All looking good. Sorry for the late response.

**Review by oomek — APPROVED** [review]

**Comment by substring** [comment]
LGTM

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (eXpl0it3r) [inline src/SFML/Window/DRM/DRMContext.cpp:190]: Some memset you'll get to rewrite once it's merged to master 😄

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (eXpl0it3r) [inline src/SFML/Window/DRM/DRMContext.cpp:51]: What's an Fb?

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (eXpl0it3r) [inline src/SFML/Window/DRM/DRMContext.cpp:206]: What's a Crtc?

Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.cpp:206]: https://events.static.linuxfound.org/sites/events/files/slides/brezillon-drm-kms.pdf

Screenshot 2022-11-27 at 3 15 30 PM

Looks like `CrtController` is a more apt name.

Comment (substring) [reply src/SFML/Window/DRM/DRMContext.cpp:206]: CRTC is DRM slang, so to say. I suggest you read the kernel documentation, https://docs.kernel.org/gpu/drm-kms.html#crtc-abstraction is a good source explaining what a CRTC is. Scroll up to the top for a short presentation of DRM. There is a number of other acronyms that are related to DRM, and should always be kept to report the DRM naming

File: src/SFML/Window/DRM/DRMContext.cpp
Comment (ChrisThrasher) [reply src/SFML/Window/DRM/DRMContext.cpp:165]: I'd honestly rather keep it this way than learn the details of how `long long` behaves across C++ standards. Plus it's nice to avoid adding casts.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by eXpl0it3r — APPROVED** [review]

**Pull request merged**
